src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,591 @@
+                        continue
+
+        # Step 30
+
+        # TH sends a Read Request Message to the DUT to read a non global attribute from all clusters at all Endpoints
+        # AttributePath = [[Attribute = Specific Non Global Attribute]] +
+
+        # On the TH verify that the DUT sends an error message and not the value of the attribute.
+        self.print_step(30, "Send the Read Request Message to the DUT to read a non global attribute from all clusters at all Endpoints")
+        found_non_global = False
+        for endpoint_id, endpoint in self.endpoints.items():
+            if not found_non_global:
+                # global_attribute_ids.AttributeIdType.kStandardNonGlobal seems to be non-existent in chip-all-clusters-app
+                # But kTest does exist -> Clusters.Objects.UnitTesting
+                for cluster_type, cluster in endpoint.items():
+
+                    if global_attribute_ids.cluster_id_type(cluster_type.id) != global_attribute_ids.ClusterIdType.kStandard:
+                        # if global_attribute_ids.cluster_id_type(cluster_type.id) == global_attribute_ids.AttributeIdType.kStandardNonGlobal:
+                        found_non_global = True
+                        continue
+                    # Temporarily commented to avoid linter errors -- will revert once output value is known so that assertion can be used correctly
+                    # non_global_cluster = cluster_type
+                    # read_request = await self.default_controller.ReadAttribute(self.dut_node_id, [(0, non_global_cluster)])
+                    # Seems to return {}?
+        # Step 31
+
+        # TH should have access to only a single cluster at one Endpoint1.
+        # TH sends a Read Request Message to the DUT to read all attributes from all clusters at Endpoint1
+        # AttributePath = [[Endpoint = Specific Endpoint]] +
+
+        # Verify that the DUT sends back data of all attributes only from that one cluster to which it has access.
+        # Verify that there are no errors sent back for attributes the TH has no access to.
+        self.print_step(31, "Send the Read Request Message to the DUT to read all attributes from all clusters at Endpoint1")
+        read_request = await self.default_controller.ReadAttribute(self.dut_node_id, [(1, descriptor_obj)])
+        asserts.assert_true(1 in read_request, "Endpoint 1 missing in response")
+        asserts.assert_true(descriptor_obj in read_request[1], "Clusters.Objects.Descriptor not in response")
+
+        # Step 32
+
+        # TH sends a Read Request Message to read all events and attributes from the DUT.
+
+        # Verify that the DUT sends back data of all attributes and events that the TH has access to.
+
+        # Clusters.Descriptor.Events doesn't seem to exist, despite https://project-chip.github.io/connectedhomeip-doc/testing/python.html#events suggesting it should
```

Comment:
Use Read here, and specify wildcard for both the attribute and events. Descriptor cluster doesn't factor in.

Reply:
Changed to use Read and added assertions to see if read_request has parameters